File: Perl/2015/day25.pl
#!/usr/bin/env perl 
#===============================================================================
#
#         FILE: day25.pl
#
#        USAGE: ./day25.pl  
#
#  DESCRIPTION: Day 25 - Santa's weather machine.
#
#      OPTIONS: ---
# REQUIREMENTS: ---
#         BUGS: ---
#        NOTES: ---
#       AUTHOR: Mark T. Johnson (mtj), markj@matzsoft.com
# ORGANIZATION: MATZ Software & Consulting
#      VERSION: 1.0
#      CREATED: 12/24/15 21:10:05
#     REVISION: ---
#===============================================================================

use strict;
use warnings;
use utf8;
use 5.010;


my $start_value = 20151125;
my $multiplier  = 252533;
my $divisor     = 33554393;
my $data        = do { local $/; $_ = <DATA>; s/[,.\n]//g; $_ };
my @words       = split q{ }, $data;
my ( $target_row, $target_col ) = map { $_ - 1 } @words[ 15, 17 ];
my ( $row, $col, $maxrow ) = ( 0, 0, 0 );
my $current = $start_value;

while ( $row != $target_row or $col != $target_col ) {
    $current = ( $current * $multiplier ) % $divisor;

    if ( $row == 0 ) {
        $row = ++$maxrow;
        $col = 0;

    } else {
        $row--;
        $col++;
    }
}

say $current;



__END__
To continue, please consult the code grid in the manual.  Enter the code at row 2947, column 3029.

File: 2015/day25.swift
//
//         FILE: main.swift
//  DESCRIPTION: day25 - Let It Snow
//        NOTES: There is a cycle in the sequence, but it takes longer to find the cycle
//               than just going all the way through.
//       AUTHOR: Mark T. Johnson, markj@matzsoft.com
//    COPYRIGHT: © 2021 MATZ Software & Consulting. All rights reserved.
//      VERSION: 1.0
//      CREATED: 07/18/21 23:37:12
//

import Foundation

let startValue = 20151125
let multiplier = 252533
let divisor    = 33554393


func parse( input: AOCinput ) -> ( Int, Int ) {
    let words = input.line.split( whereSeparator: { " ,.".contains( $0 ) } )
    let indices = words.indices.filter { Int( words[$0] ) != nil }
    let dict = Dictionary( uniqueKeysWithValues: indices.map { ( words[$0-1], Int( words[$0] )! ) } )
    
    return ( dict["row"]! - 1, dict["column"]! - 1 )
}


func part1( input: AOCinput ) -> String {
    let ( targetRow, targetCol ) = parse( input: input )
    let diagonal = targetRow + targetCol
    let repeatCount = diagonal * ( diagonal + 1 ) / 2 + targetCol
    var current = startValue
    
    for _ in 0 ..< repeatCount {
        current = ( current * multiplier ) % divisor
    }
    return "\(current)"
}


func part2( input: AOCinput ) -> String {
    return "None"
}


try runTests( part1: part1 )
try runTests( part2: part2 )
try solve( part1: part1 )
try solve( part2: part2 )
